### DISCUSSION TITLE
GSOC 2021 TO-DOs

### DISCUSSION THREAD

**Discussion by ilyaZar** [rcppsmc/rcppsmc#57] [Summer of Code]
This is to keep track of all the TO-DOs, starting from the projects proposal and adding subsequently all other issues that arise during the project time.

# Main To-Dos

## Tracking of ancestral lines

**_Aim:_**
Modify underlying C++ code to allow ancestral lineages to be traced. 

- [x] switch to ancestor lineage tracking (e.g. via a new `HistoryType::AL`)
- [x] ancestor indices to be saved in a suitable container in the following ways:
  - [x] a standard container `std::vector<>`
  - [ ] a more efficient tree-based approach
- [x] add getters/setters to ancestral lines
- [x] add member functions to compute ancestral lines
- [x] example use case for SV model 
- [x] deliver graphical illustrations of the ancestral lines of the particle system

## Conditional SMC

**_Aim_**: Provide additional template support for conditional SMC algorithms

- [x] add a derived `conditionalSampler` class that inherits from the main sampler class to support reference trajectories
  - [x] implement a conditioning step
  - [x] implement correct weighting of the conditional reference value
  - [x] the conditioning step successfully resets the particle coordinates of the trajectory to be retained to the corresponding
  coordinate value of the reference trajectory
- [x] change resampling facilities of base `sampler` class that implement conditional resampling
  - [x] add conditional multinomial resampling
  - [x] add conditional stratified resampling
  - [x] add conditional systematic resampling
  - [ ] add conditional residual resampling 
- [ ] example use case: Particle Gibbs sampler for the SV model
  - [ ] graphical illustration: frozen path issue
  - [ ] compute and plot update rates

## Backward sampling

**_Aim_**: 
Include support for backward/ancestor sampling by making use of these ancestor indices and conditional SMC algorithms.

- [ ] Allows the user to implement a backwards sampling step (interface similar to the virtual functions for user defined proposal and weights)
- [ ] example use case: Particle Gibbs sampler with ancestor sampling for the SV model
  - [ ] graphical illustration: break up of frozen path trajectory
  - [ ] computing and plotting update rates as an illustration of the mixing properties of the sampler that now should improve
  considerably

# Additional To-Dos

- [ ] Major revision and update of the documentation.
- [ ] Provide consistent interfaces for `sampler` and `conditionalSampler` for the printing facilities (question: number of digits to print belongs within an individual sampler object or more globally? -> in any case consistent between both classes)
- [ ] Fix redundant `static_cast<int>` leaving type `long` where possible
- [ ] Decide on use of typedefs instead explicit naming (`arma::Col<unsigned int>` vs. `arma::icovec`)
- [ ] Estimation of particle filter variance using the kind of techniques developed e.g. in Leeand Whiteley (2018).
- [ ] Illustration of nonlinear state space smoothing using the conditional particle filtersee e.g. Svensson, Schön, and Kok (2015).

**Comment by adamjohansen** [comment]
Thanks, it's very good to have these collected somewhere. (Of course there's no reason to suppose that everything that appears on this list needs to get done as part of this GSoC project... but it is definitely good to have a list of the things that we know need doing and we can transfer things from this list to standard issues later if that's sensible.)

**Comment by LeahPrice** [comment]
Thanks, this is a nice and comprehensive list!

**Comment by ilyaZar** [comment]
Hi. I think I'll post here the link to this years GSoC https://ilyazar.github.io/gsoc2021/

This topic is certainly not the best fit, but I don't really want to open another just for a link, and in some sense, the webpage does describe parts above checked DONEs of the TO-DO list :)
